Update gimli symbolizer to return inlined functions

philipc opened the pull request:
The addr2line crate can now return inlined functions. It also requires us to load the file ourselves now.

There doesn't appear to be any tests for inlined functions. I hacked the smoke test to try it out, but haven't included that in this PR since I wasn't sure how well other symbolizers would handle that.

cc @fitzgen

alexcrichton commented:
Thanks for the PR! Would it be possible to trim down the deps a bit? I'd be ok with a little unsafety here, there's already plenty elsewhere in the crate!

fitzgen commented on src/symbolize/gimli.rs:
```diff
@@ -109,33 +151,46 @@ pub fn resolve(addr: *mut c_void, cb: &mut FnMut(&super::Symbol)) {
+            }
+        }
+
+        if !found_sym {
```
Nice. Maybe leave a comment about the symbol table fallback here?

fitzgen commented on src/symbolize/gimli.rs:
```diff
@@ -109,33 +151,46 @@ pub fn resolve(addr: *mut c_void, cb: &mut FnMut(&super::Symbol)) {
 pub struct Symbol {
     addr: usize,
-    file: PathBuf,
+    file: Option<PathBuf>,
```
This is optional now because the fallback might not have this info, right?

philipc replied:
Both that and because addr2line returns an option too, for the unlikely case where we found a function name, but not location info.

fitzgen commented on src/lib.rs:
```diff
@@ -96,6 +96,27 @@ extern crate addr2line;
+#[cfg(all(feature = "gimli-symbolize",
+          unix,
+          target_os = "linux"))]
+extern crate stable_deref_trait;
```
Does this crate already depend on `cfg-if`? If so, maybe we could do

```rust
cfg_if! {
    if #[cfg(all(feature = "gimli-symbolize", unix, target_os = "linux"))] {
        extern crate addr2line;
        extern crate findshlibs;
        extern crate gimli;
        extern crate memmap;
        extern crate object;
        extern crate rental;
        extern crate stable_deref_trait;
    } else {
        // nothing here, but cfg-if! requires an else
    }
}
```

Which is less `#[cfg(..)]` noise.

philipc replied:
Done, but without the else because cfg-if doesn't require it anymore.

fitzgen commented:
> Thanks for the PR! Would it be possible to trim down the deps a bit? I'd be ok with a little unsafety here, there's already plenty elsewhere in the crate!

Are you specifically referring to `rental` and `stable_deref_trait`?

It isn't clear to me based on this comment what changes are necessary for merging this PR.

alexcrichton commented:
Er sorry yeah, if this could avoid rental/stable_deref_trait that'd be great (I already only have a passing understanding of what's going on). Beyond that though looks good!

philipc commented:
I've removed the rental/stable_deref_trait deps and addressed the other comments. Subverting the borrow checker isn't something I'm used to, so please check the mem::transmute usage.

philipc commented:
The use of  '?' for Option broke the rustc 1.21 build. Should I avoid that or bump the version?

alexcrichton commented:
Oh oops! Nah I think it's fine to just bump the required version
